### PR TITLE
Update DefaultResourceHandlerFactory.cs

### PR DESCRIPTION
**Pull request by lamerbug**
- using ConcurrentDictionary
- create canonical URL from any string
- update value if key exists
- return bool result when register and unregister

### CODE REVIEW

File: CefSharp/DefaultResourceHandlerFactory.cs [hunk 1]
```diff
@@ -3,27 +3,34 @@
...
+        public virtual bool RegisterHandler(string url, IResourceHandler handler)
         {
-            Handlers[url] = handler;
+            Uri uri;
+            if (Uri.TryCreate(url, UriKind.Absolute, out uri) || Uri.TryCreate("http://" + url, UriKind.Absolute, out uri))
+            {
+                Handlers.AddOrUpdate(uri.ToString(), handler, (k, v) => handler);
+                return true;
+            }
+                return false;
```
Comment (amaitland) [inline CefSharp/DefaultResourceHandlerFactory.cs]: Minor formatting issue.

File: CefSharp/DefaultResourceHandlerFactory.cs [hunk 2]
```diff
@@ -3,27 +3,34 @@
...
-            Handlers = new Dictionary<string, IResourceHandler>(comparer ?? StringComparer.OrdinalIgnoreCase);
+            Handlers = new ConcurrentDictionary<string, IResourceHandler>(comparer ?? StringComparer.OrdinalIgnoreCase);
         }
 
-        public virtual void RegisterHandler(string url, IResourceHandler handler)
+        public virtual bool RegisterHandler(string url, IResourceHandler handler)
         {
-            Handlers[url] = handler;
+            Uri uri;
+            if (Uri.TryCreate(url, UriKind.Absolute, out uri) || Uri.TryCreate("http://" + url, UriKind.Absolute, out uri))
```
Comment (amaitland) [inline CefSharp/DefaultResourceHandlerFactory.cs]: In this case if they don't provide a `scheme name` then I'd rather throw an exception than implicitly add `http`.

**Comment by amaitland** [comment]
Fixes #1106

**Comment by lamerbug** [comment]
>>In this case if they don't provide a scheme name then I'd rather throw an exception than implicitly add http.

I think it's ok, That's a default behaviour of UriBuilder, and browser does the same. So user can get working any string. (I have a use case when user allowed to manually register it)

**Comment by rassilon** [comment]
> No, I think it's ok, That's a default behaviour of UriBuilder, and browser does the same. So user can get working any string. (I have a use case when user allowed to manually register it)

I think I'm with @amaitland here, if you'd like to do something specific for "urls" without schemes, I would suggest doing so at the application level. Sort of like Chrome/IE change the request into a google/bing/whatever search for such URLs if its not obviously domain name like.

So, I think you'd have to explain to us a bit more why CEFSharp prevents you from doing something like that at the application level.  My gut instinct is to infer that while your application might be able to be tweaked to do such a thing, that for various other reasons its just cheaper/less risk/etc... to make this change in CEFSharp.

The ConcurrentDictionary part is just fine of course. I've even commented about needing to getting around to doing it in the past.

**Comment by lamerbug** [comment]
>So, I think you'd have to explain to us a bit more why CEFSharp prevents you from doing something like that at the application level.

Forgot to mention that while resolving #1106 I just was implementing custom ResourceHandlerFactory so of course any changes to DefaultResourceHandlerFactory is a subject to review from your side.

> that for various other reasons its just cheaper/less risk/etc... to make this change in CEFSharp.

We have to canonize url anyway to get it working. If there would be a choice to do so or not, then of course we would move all the logic to application level.

**Comment by amaitland** [comment]
> We have to canonize url anyway to get it working. If there would be a choice to do so or not, then of course we would move all the logic to application level.

We'll keep that change of course :+1:

**Pull request merged**